In [1]:

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import plotly.express as px # Data vizualization
import plotly.graph_objects as go




In [2]:
# Read used car data from a CSV file, drop the 'index' column, and store the resulting DataFrame in a variable
used_car = pd.read_csv(r"merged-data.csv").drop("index", axis=1)
used_car_data = used_car.copy() # Create a copy of the used car DataFrame

#                                        Columns

| Column name        | Description                                                                |
|-------------------|----------------------------------------------------------------------------|
| dateCrawled       | Date the car was crawled. (Date)                                           |
| name              | Name of the car. (String)                                                  |
| seller            | Type of seller (private or dealer). (String)                               |
| offerType         | Type of offer (e.g. sale, repair, etc.). (String)                         |
| price             | Price of the car. (Integer)                                                |
| abtest            | Test type (A or B). (String)                                              |
| vehicleType       | Type of vehicle (e.g. SUV, sedan, etc.). (String)                         |
| yearOfRegistration| Year the car was registered. (Integer)                                     |
| gearbox           | Type of gearbox (manual or automatic). (String)                           |
| powerPS           | Power of the car in PS. (Integer)                                          |
| model             | Model of the car. (String)                                                 |
| kilometer         | Kilometers the car has been driven. (Integer)                              |
| monthOfRegistration | Month the car was registered. (Integer)                                   |
| fuelType          | Type of fuel (e.g. diesel, petrol, etc.). (String)                        |
| brand             | Brand of the car. (String)                                                 |
| notRepairedDamage | Whether or not the car has any damage that has not been repaired. (String) |
| dateCreated       | Date the car was created. (Date)                                           |
| nrOfPictures      | Number of pictures of the car. (Integer)                                   |
| postalCode        | Postal code of the car. (Integer)                                          |
| lastSeen          | Date the car was last seen. (Date)                                         |


In [3]:
# Display the first 10 rows of the used car DataFrame
used_car_data.head(10)

dateCrawled                                               name  \
0  2016-03-24 11:52:17                                         Golf_3_1.6   
1  2016-03-24 10:58:45                               A5_Sportback_2.7_Tdi   
2  2016-03-14 12:52:21                     Jeep_Grand_Cherokee_"Overland"   
3  2016-03-17 16:54:04                                 GOLF_4_1_4__3TÜRER   
4  2016-03-31 17:25:20                     Skoda_Fabia_1.4_TDI_PD_Classic   
5  2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
6  2016-04-01 20:48:51                        Peugeot_206_CC_110_Platinum   
7  2016-03-21 18:54:38                       VW_Derby_Bj_80__Scheunenfund   
8  2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   
9  2016-03-17 10:53:50  VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...   

   seller offerType  price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot    480     test         NaN                1993   
1  privat   Angebot  18300     test       coupe                2011   
2  privat   Angebot   9800     test         suv                2004   
3  privat   Angebot   1500     test  kleinwagen                2001   
4  privat   Angebot   3600     test  kleinwagen                2008   
5  privat   Angebot    650     test   limousine                1995   
6  privat   Angebot   2200     test      cabrio                2004   
7  privat   Angebot      0     test   limousine                1980   
8  privat   Angebot  14500  control         bus                2014   
9  privat   Angebot    999     test  kleinwagen                1998   

     gearbox  powerPS    model  kilometer  monthOfRegistration fuelType  \
0    manuell        0     golf     150000                    0   benzin   
1    manuell      190      NaN     125000                    5   diesel   
2  automatik      163    grand     125000                    8   diesel   
3    manuell       75     golf     150000                    6   benzin   
4    manuell       69    fabia      90000                    7   diesel   
5    manuell      102      3er     150000                   10   benzin   
6    manuell      109  2_reihe     150000                    8   benzin   
7    manuell       50   andere      40000                    7   benzin   
8    manuell      125    c_max      30000                    8   benzin   
9    manuell      101     golf     150000                    0      NaN   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0  volkswagen               NaN  2016-03-24 00:00:00             0   
1        audi                ja  2016-03-24 00:00:00             0   
2        jeep               NaN  2016-03-14 00:00:00             0   
3  volkswagen              nein  2016-03-17 00:00:00             0   
4       skoda              nein  2016-03-31 00:00:00             0   
5         bmw                ja  2016-04-04 00:00:00             0   
6     peugeot              nein  2016-04-01 00:00:00             0   
7  volkswagen              nein  2016-03-21 00:00:00             0   
8        ford               NaN  2016-04-04 00:00:00             0   
9  volkswagen               NaN  2016-03-17 00:00:00             0   

   postalCode             lastSeen  
0       70435  2016-04-07 03:16:57  
1       66954  2016-04-07 01:46:50  
2       90480  2016-04-05 12:47:46  
3       91074  2016-03-17 17:40:17  
4       60437  2016-04-06 10:17:21  
5       33775  2016-04-06 19:17:07  
6       67112  2016-04-05 18:18:39  
7       19348  2016-03-25 16:47:58  
8       94505  2016-04-04 23:42:13  
9       27472  2016-03-31 17:17:06

In [4]:
# Display the data types of each column in the used car DataFrame
used_car_data.dtypes

dateCrawled            object
name                   object
seller                 object
offerType              object
price                   int64
abtest                 object
vehicleType            object
yearOfRegistration      int64
gearbox                object
powerPS                 int64
model                  object
kilometer               int64
monthOfRegistration     int64
fuelType               object
brand                  object
notRepairedDamage      object
dateCreated            object
nrOfPictures            int64
postalCode              int64
lastSeen               object
dtype: object

#### The three columns dateCrawled, dateCreated, and lastSeen in the used_car_data DataFrame are currently being stored as object data types, which means that they are being treated as strings. However, these columns likely contain date and time information, and it would be more useful to store them as datetime data types.

#### Storing dates and times as datetime data types allows you to perform operations such as subtraction and comparison on them, as well as formatting them for display. It also allows you to extract the year, month, day, hour, etc. from the datetime value if needed.

#### To convert these columns to datetime data types, you can use the pd.to_datetime function from the pandas library.

In [5]:
used_car_data['dateCrawled'] = pd.to_datetime(used_car_data['dateCrawled']) # Convert the 'dateCrawled' column to datetime data type
used_car_data['dateCreated'] = pd.to_datetime(used_car_data['dateCreated']) # Convert the 'dateCreated' column to datetime data type
used_car_data['lastSeen'] = pd.to_datetime(used_car_data['lastSeen']) # Convert the 'lastSeen' column to datetime data type

In [6]:
# Display the data types of each column in the used car DataFrame
used_car_data.dtypes

dateCrawled            datetime64[ns]
name                           object
seller                         object
offerType                      object
price                           int64
abtest                         object
vehicleType                    object
yearOfRegistration              int64
gearbox                        object
powerPS                         int64
model                          object
kilometer                       int64
monthOfRegistration             int64
fuelType                       object
brand                          object
notRepairedDamage              object
dateCreated            datetime64[ns]
nrOfPictures                    int64
postalCode                      int64
lastSeen               datetime64[ns]
dtype: object

In [7]:
# Display summary statistics for numerical columsn in the used car DataFrame
used_car_data.describe()

dateCrawled         price  yearOfRegistration  \
count                         371528  3.715280e+05       371528.000000   
mean   2016-03-21 13:30:31.202717184  1.729514e+04         2004.577997   
min              2016-03-05 14:06:22  0.000000e+00         1000.000000   
25%       2016-03-13 12:51:46.500000  1.150000e+03         1999.000000   
50%              2016-03-21 18:36:22  2.950000e+03         2003.000000   
75%    2016-03-29 14:49:31.249999872  7.200000e+03         2008.000000   
max              2016-04-07 14:36:58  2.147484e+09         9999.000000   
std                              NaN  3.587954e+06           92.866598   

             powerPS      kilometer  monthOfRegistration  \
count  371528.000000  371528.000000        371528.000000   
mean      115.549477  125618.688228             5.734445   
min         0.000000    5000.000000             0.000000   
25%        70.000000  125000.000000             3.000000   
50%       105.000000  150000.000000             6.000000   
75%       150.000000  150000.000000             9.000000   
max     20000.000000  150000.000000            12.000000   
std       192.139578   40112.337051             3.712412   

                         dateCreated  nrOfPictures    postalCode  \
count                         371528      371528.0  371528.00000   
mean   2016-03-20 19:42:51.495015168           0.0   50820.66764   
min              2014-03-10 00:00:00           0.0    1067.00000   
25%              2016-03-13 00:00:00           0.0   30459.00000   
50%              2016-03-21 00:00:00           0.0   49610.00000   
75%              2016-03-29 00:00:00           0.0   71546.00000   
max              2016-04-07 00:00:00           0.0   99998.00000   
std                              NaN           0.0   25799.08247   

                            lastSeen  
count                         371528  
mean   2016-03-30 03:37:50.548152832  
min              2016-03-05 14:15:08  
25%       2016-03-23 11:32:10.500000  
50%              2016-04-03 23:17:15  
75%              2016-04-06 10:45:02  
max              2016-04-07 14:58:51  
std                              NaN

In [8]:
# Calculate the percentage of missing values in each column of the used car DataFrame
persentage_of_missing = used_car_data.isnull().sum().sort_values(ascending=False) * 100 / len(used_car_data)
persentage_of_missing

notRepairedDamage      19.395577
vehicleType            10.192771
fuelType                8.986133
model                   5.513447
gearbox                 5.439429
kilometer               0.000000
postalCode              0.000000
nrOfPictures            0.000000
dateCreated             0.000000
brand                   0.000000
monthOfRegistration     0.000000
dateCrawled             0.000000
name                    0.000000
powerPS                 0.000000
yearOfRegistration      0.000000
abtest                  0.000000
price                   0.000000
offerType               0.000000
seller                  0.000000
lastSeen                0.000000
dtype: float64

#### This table shows the percentage of missing values in each column of the used car data. The columns are sorted in descending order by the percentage of missing values.

#### The notRepairedDamage column has the highest percentage of missing values (19.4%), followed by the vehicleType column (10.2%), the fuelType column (9.0%), the model column (5.5%), and the gearbox column (5.4%). These columns have a relatively large number of missing values compared to the other columns.

#### On the other hand, the kilometer, postalCode, nrOfPictures, dateCreated, brand, monthOfRegistration, dateCrawled, name, powerPS, yearOfRegistration, abtest, price, offerType, seller, and lastSeen columns have 0% missing values. This means that all values in these columns are present, and there are no missing values.

#### It is generally important to identify and handle missing values in a dataset, as missing values can affect the accuracy and reliability of analysis and modeling. In some cases, missing values may need to be imputed or estimated, while in other cases they may need to be dropped or replaced with a default value. The appropriate approach depends on the context and purpose of the data analysis.

In [9]:
# Create a bar chart showing the percentage of missing values in each column of the used car DataFrame
fig = px.bar(x=persentage_of_missing.index, y=persentage_of_missing, title= "Percentage of Missing Values in Used Car Data", labels={"x" : "Columns", "y" : "Percentage of Missing Values"})
fig.update_traces(marker_color = "red") # Update the color of the bars in the bar chart to red
# fig.show() # Display the bar chart

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

Figure({
    'data': [{'alignmentgroup': 'True',
              'hovertemplate': 'Columns=%{x}<br>Percentage of Missing Values=%{y}<extra></extra>',
              'legendgroup': '',
              'marker': {'color': 'red', 'pattern': {'shape': ''}},
              'name': '',
              'offsetgroup': '',
              'orientation': 'v',
              'showlegend': False,
              'textposition': 'auto',
              'type': 'bar',
              'x': array(['notRepairedDamage', 'vehicleType', 'fuelType', 'model', 'gearbox',
                          'kilometer', 'postalCode', 'nrOfPictures', 'dateCreated', 'brand',
                          'monthOfRegistration', 'dateCrawled', 'name', 'powerPS',
                          'yearOfRegistration', 'abtest', 'price', 'offerType', 'seller',
                          'lastSeen'], dtype=object),
              'xaxis': 'x',
              'y': array([19.39557718, 10.19277147,  8.98613294,  5.51344717,  5.43942852,
                           0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
                           0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
                           0.        ,  0.        ,  0.        ,  0.        ,  0.        ]),
              'yaxis': 'y'}],
    'layout': {'barmode': 'relative',
               'legend': {'tracegroupgap': 0},
               'template': '...',
               'title': {'text': 'Percentage of Missing Values in Used Car Data'},
               'xaxis': {'anchor': 'y', 'domain': [0.0, 1.0], 'title': {'text': 'Columns'}},
               'yaxis': {'anchor': 'x', 'domain': [0.0, 1.0], 'title': {'text': 'Percentage of Missing Values'}}}
})

In [10]:
most_common = used_car_data["notRepairedDamage"].value_counts(ascending=False).index[0] # Find the most common value in the 'notRepairedDamage' column and assign it to the variable 'most_common'
used_car_data["notRepairedDamage"].fillna(most_common, inplace=True) # Replace missing values in notRepairedDamage column with most common value

/var/folders/v0/xjnkd3c1787_ttkzjw009h8c0000gn/T/ipykernel_59669/2763428692.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [11]:
most_common = used_car_data["vehicleType"].value_counts(ascending=False).index[0] # Find the most common value in the 'vehicleType' column and assign it to the variable 'most_common'
used_car_data["vehicleType"].fillna(most_common, inplace=True) # Replace missing values in vehicleType column with most common value

/var/folders/v0/xjnkd3c1787_ttkzjw009h8c0000gn/T/ipykernel_59669/645338331.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [12]:
most_common = used_car_data["fuelType"].value_counts(ascending=False).index[0] # Find the most common value in the 'fuelType' column and assign it to the variable 'most_common'
used_car_data["fuelType"].fillna(most_common, inplace=True) # Replace missing values in vehicleType column with most common value

/var/folders/v0/xjnkd3c1787_ttkzjw009h8c0000gn/T/ipykernel_59669/4051799966.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [13]:
most_common = used_car_data["gearbox"].value_counts(ascending=False).index[0] # Find the most common value in the 'gearbox' column and assign it to the variable 'most_common'
used_car_data["gearbox"].fillna(most_common, inplace=True) # Replace missing values in gearbox column with most common value

/var/folders/v0/xjnkd3c1787_ttkzjw009h8c0000gn/T/ipykernel_59669/1076775120.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [14]:
# We can extract model name from name column but we are not interesting right now.
used_car_data["model"].fillna("unknown", inplace=True)

/var/folders/v0/xjnkd3c1787_ttkzjw009h8c0000gn/T/ipykernel_59669/3418289010.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [15]:
useless_col = ["nrOfPictures", "lastSeen", "dateCrawled", "name", "monthOfRegistration", "dateCreated", "postalCode", "seller", "offerType"] # Identify columns that are not useful for analysis and store them in a list called 'useless_col
new_used_car = used_car_data.drop(useless_col, axis=1) # Drop the unnecessary columns from the 'used_car_data' dataframe and store the resulting dataframe in 'new_used_car'

#### Outliers are data points that are significantly different from the rest of the data. These data points can sometimes have a significant effect on statistical analyses and may lead to incorrect conclusions. Therefore, it is often a good practice to identify and remove outliers from the data before performing any statistical analysis. This function uses the interquartile range (IQR) method to identify and remove outliers from a dataframe. The IQR is calculated as the difference between the first and third quartiles (Q1 and Q3). Data points that are less than Q1 - 1.5 * IQR or m than Q3 + 1.5 * IQR are considered outliers and are removed from the dataframe.

In [16]:
def remove_outliers(df: pd.DataFrame) -> pd.DataFrame:
    """
    This function removes outliers from a given dataframe by calculating the lower and upper bounds of the interquartile range (IQR). 
    Outliers are defined as values that are less than Q1 - 1.5 * IQR or greater than Q3 + 1.5 * IQR, where Q1 and Q3 are the first and third quartiles,
    respectively.

    Input:
    df (pd.DataFrame): A dataframe containing numeric columns

    Output:
    df (pd.DataFrame): A dataframe containing only rows that are not considered outliers
    """
    
    # Select only numeric columns
    numeric_df = df.select_dtypes(include=['number'])

    Q1 = numeric_df.quantile(0.1) # find the Q1 value for each numeric column
    Q3 = numeric_df.quantile(0.9) # find the Q3 value for each numeric column
    IQR = Q3 - Q1 # find the interquartile range for each numeric column
    
    # return the dataframe with rows that do not have any values lower than Q1 - 1.5 * IQR or higher than Q3 + 1.5 * IQR
    return df[~((numeric_df < (Q1 - 1.5 * IQR)) | (numeric_df > (Q3 + 1.5 * IQR))).any(axis=1)]

clean_user_car = remove_outliers(new_used_car)



In [17]:
from sklearn.preprocessing import LabelEncoder

#### Based on the correlation matrix, it appears that there are several variables that are significantly correlated with the price of the car. The year of registration and power in PS are positively correlated with price, while the kilometer driven is negatively correlated with price. Model of the car, as well as the type of fuel and gearbox, also seem to have some impact on the price. It is worth noting that these correlations may not necessarily indicate a causal relationship, but rather a statistical association between the variables.

In [18]:
from joblib import dump
encoded_user_car = clean_user_car.copy() # Make a copy of the clean_user_car dataframe and assign it to encoded_user_car
cat_cols = ["notRepairedDamage", "vehicleType", "abtest", "gearbox", "model", "fuelType", "brand"]
encoders = {}

for col in cat_cols:
    encoder = LabelEncoder()
    encoded_user_car[col] = encoder.fit_transform(encoded_user_car[col])
    encoders[col] = encoder

for col, encoder in encoders.items():
    dump(encoder, f'{col}_encoder.joblib')

# fig = go.Figure() # Initialize a Figure object from the plotly library
# fig.add_trace(
#     go.Heatmap(
#         x = corr.index,
#         y = corr.index,
#         z = corr,
#         text=corr.values,
#         texttemplate='%{text:.2f}'
#     )
# ) # Add a Heatmap trace to the figure
# fig.update_layout(width = 1000, height = 1000, title = "Heatmap showing correlations between variables in the used car dataset") # Update the layout of the figure to set the width and height
# fig.show() #Show the heatmap

# EDA

### What is the relationship between the price of a car and whether or not it has damage that has not been repaired?

#### According to the plot, cars that have not been damaged or have already been repaired tend to have higher prices compared to cars that have not yet been repaired or have been damaged. This could potentially be because cars that are in better condition tend to be more desirable and therefore command higher prices. Additionally, it could also be because the cost of repairing damage to a car can be significant, so buyers may be willing to pay more for a car that is already in good condition to avoid having to make those repairs themselves.

In [19]:
# Create a scatter plot showing the relationship between price and not repaired damage, with the color of the points indicating the not repaired damage status
fig = px.scatter(data_frame=clean_user_car, x = "price", color="notRepairedDamage", title="Relationship Between Price and Not Repaired Damage")
# fig.show() # Show the plot

### What is the relationship between powerPS and price in the used car data?

#### Based on the plot, it appears that there is a slight positive relationship between the power of the car in PS and the price of the car. This means that, as the power of the car increases, the price of the car tends to increase as well. However, it is important to note that this relationship may not be particularly strong, and other factors may also be influencing the price of the car.

In [20]:
fig = px.scatter(data_frame=clean_user_car.iloc[:10000], x = "powerPS", y = "price", title="Relationship Between PowerPS and Price") # Plot scatter plot of price vs. powerPS in the clean_user_car dataframe, using the first 10000 rows
# fig.show() # Show the plot

In [21]:
fig = px.scatter(data_frame=clean_user_car.iloc[:10000], x = "kilometer", y = "price", title="Relationship Between Kilometer and Price") # Plot scatter plot of price vs. Kilometer in the clean_user_car dataframe, using the first 10000 rows
# fig.show() # Show the plot

### What is the relationship between price and gearbox type in used cars?

#### Based on this plot, it appears that cars with automatic gearboxes tend to be slightly more expensive than cars with manual gearboxes. This can potentially be due to various factors such as the popularity and demand for automatic gearboxes, or perhaps the additional features and technology that may come with cars equipped with automatic gearboxes.

In [22]:
fig = px.box(data_frame=clean_user_car, x = "price", color="gearbox", title="Price Distribution by Gearbox Type")
# fig.show()

### What are the three most expensive car brands based on their mean price?

#### According to this plot, the top three most expensive car brands are Porsche, Land Rover, and Mini, in that order. The mean prices for these brands are 16840.658348, 12011.764706, and 9805.431529, respectively. This means that, on average, Porsche cars are the most expensive, followed by Land Rover, and then Mini. The sum and count values indicate the sum of price and number of cars for each brand, respectively, and do not directly contribute to determining the most expensive car brands.

In [23]:
expensive_cars = clean_user_car["price"].groupby(clean_user_car["brand"]).agg(["sum", "count", "mean"]).sort_values(by = "mean", ascending=False)
fig = px.bar(data_frame=expensive_cars, y = "mean", color=expensive_cars.index)
# fig.show()

# Modelling

In [24]:
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestRegressor, ExtraTreesRegressor
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, make_scorer
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeRegressor
import xgboost as xgb

In [25]:
X = encoded_user_car.drop("price", axis = 1).values
y = encoded_user_car["price"].values

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15)
print(f"Train Shape : {X_train.shape} --------- Test Shape : {X_test.shape}")

Train Shape : (292123, 10) --------- Test Shape : (73031, 10)


In [27]:
scores = {"r2_score" : make_scorer(r2_score), "mae" : make_scorer(mean_absolute_error), "mse" : make_scorer(mean_squared_error)}

In [28]:
model = xgb.XGBRegressor(tree_method="hist", n_estimators = 1000)
result = cross_validate(model, X_train, y_train, cv = 5, verbose = 2, scoring=scores)

[CV] END .................................................... total time=   6.0s
[CV] END .................................................... total time=   6.1s
[CV] END .................................................... total time=   6.7s
[CV] END .................................................... total time=   6.4s
[CV] END .................................................... total time=   7.4s


In [29]:
for key, value in result.items():
    print(f"{key} : {np.mean(value)} +- {np.std(value)}")

fit_time : 6.195264053344727 +- 0.5062116151810824
score_time : 0.308117151260376 +- 0.005784157803178161
test_r2_score : 0.8779118418693542 +- 0.00308739971753654
test_mae : 1143.5289016786987 +- 5.569579748303449
test_mse : 4246271.700010474 +- 117822.23120596055


In [30]:
model.fit(X_train, y_train)
pred = model.predict(X_test)
print(f"R2 Score : {r2_score(y_test, pred)}")

R2 Score : 0.8818391561508179


In [31]:
from joblib import dump

# Assuming label_encoder is your LabelEncoder instance
dump(encoder, 'label_encoder.joblib')
# Assuming model is your trained machine learning model
dump(model, 'latest_model.joblib')


['latest_model.joblib']